### ISSUE TITLE
Expo support

### ISSUE THREAD

**Issue by shirakaba** [microsoft/react-native-windows#13534]
### Summary

Expo doesn't officially support React Native Windows, so l'd like to track what's missing.

### Motivation

Meta now officially recommend using React Native via a framework such as Expo. A pain-point to adopting out-of-tree platforms is setting up all the boilerplate, and Expo has an excellent template system for taming all of that, furthermore enabling easy updates simply by bumping the version of the SDK and running what they call a "prebuild" again to regenerate project files.

### Open Questions

I'll knowledge-share how I got `react-native-windows` working alongside `react-native-macos` and `react-native` (all v73) on Expo SDK 50. It lacks config plugins, but you can at least use the same Expo CLI to start and bundle app projects.

## metro.config.js

TODO: I'm just going to save this post and edit it, because Safari has already crashed once whilst filing this issue

**Comment by kziemski** [comment]
@shirakaba ty for this.

**Comment by jonthysell** [comment]
@shirakaba Thank you for this.

With the clear upstream guidance of "just use expo" I'm actually preparing an internal presentation on the current state of expo to present to the RNW team so we can maybe fill some gaps.

I'm not sure how expo prebuild actually works under the covers, but perhaps similar to what you did with `react-native autolink-windows`, we now have an `react-native init-windows` command for generating the `windows` folder (and making updates to metro.config, etc) that is template-based.

Part of a potential todo list would be to create an "expo (prebuild) friendly" template (either new or if possible by fixing the existing ones). I would love to see an example of "here's a new expo project that got at least some parts of windows working (esp wrt. metro)".

**Comment by shirakaba** [comment]
> With the clear upstream guidance of "just use expo" I'm actually preparing an internal presentation on the current state of expo to present to the RNW team so we can maybe fill some gaps.

Sounds exciting!

> I'm not sure how expo prebuild actually works under the covers, but perhaps similar to what you did with react-native autolink-windows, we now have an `react-native init-windows` command for generating the windows folder (and making updates to metro.config, etc) that is template-based.

Ah yes, I've used that. Could you show me where the template lives?

> Part of a potential todo list would be to create an "expo (prebuild) friendly" template (either new or if possible by fixing the existing ones). I would love to see an example of "here's a new expo project that got at least some parts of windows working (esp wrt. metro)".

I'm fairly familiar with Expo Prebuild now as I've been contributing some things to it to advance support for `react-native-macos` (and NativeScript, but that's another story), so I'd be happy to create a strawman template for out-of-tree platforms.

There are some obstacles to out-of-tree platform support of `expo prebuild` and `expo create` (which performs a prebuild after downloading a template):

- `react-native-macos` is lagging on v0.73, which corresponds to Expo SDK v50
- the latest Expo SDK is v51. The CLI in this version is very desirable as it allows users to [use GitHub URLs for prebuild templates](https://github.com/expo/expo/pull/26631).
- the v50 CLI corrupts any binary files stored in the template, so you wouldn't be able to safely place any `.dll` files into the template. My [PR](https://github.com/expo/expo/pull/27212) to fix that landed in SDK v51, so whenever I have templates involving binary files, I use the v50 SDK for the runtime code yet use the v51 CLI any time I want to invoke the `create` and `prebuild` commands.
- however, that same PR creates a new problem. Previously, you could just write the magic `HelloWorld` placeholder string in any file in the whole template (even a JPG) and it'd get renamed to your desired project name; but now that string-replacement only runs on an explicit list of file paths. It's a simple matter to add new paths (as I show in my not-yet-merged [PR](https://github.com/expo/expo/pull/30309) to set up the paths to be renamed on macOS projects), but it will need to be done for Windows and it can take 1-2 months for even a simple PR to get merged.
- so I'm thinking, as an interim measure to unblock usage of the Expo v51 CLI, we could possibly commit the template files with our own placeholder string `ByeWorld` and just run our own renaming script after any usages of `create` or `prebuild`.

Support for other features of the Expo CLI (like the `start` command, which seems to ensure that any config plugins have run) are another story and will require further thought.

Will see if I can create that strawman template soon to advance discussions.

**Comment by wodin** [comment]
@shirakaba since these days the `expo` CLI is a dependency of the app, I assume you could fork it in the short term while waiting for PRs to land? Then I suppose you would just have to install the fork after running `npx create-expo-app`.

When you say `expo create` are you referring to `create-expo-app`?

When you say "like the `start` command, which seems to ensure that any config plugins have run", do you mean `expo run` (which automatically runs `prebuild`)?

**Comment by shirakaba** [comment]
For sure, I can fork to my heart's content – just a bit of a harder sell when it comes to putting it into other people's hands. Every step with asterisks increases friction in usage, and we don't want to make people have to commit to using a forked CLI that I might give up on maintaining one day.

> When you say `expo create` are you referring to `create-expo-app`?

I'm referring to [create-expo](https://github.com/expo/expo/tree/main/packages/create-expo), sorry, yeah.

> When you say "like the start command, which seems to ensure that any config plugins have run", do you mean expo run (which automatically runs prebuild)?

Ah yep, I meant `expo run:<platform>` (theoretically in this case `expo run:windows`), sorry. And exactly, it runs prebuild implicitly if the platform folder is missing.

**Comment by tido64** [comment]
> For some reason, the `@rnx-kit/metro-config` recommended default config didn't work out-of-the-box for me ([microsoft/rnx-kit#3257](https://github.com/microsoft/rnx-kit/issues/3257)) so I'd love to simplify this.

FYI, as of [1.3.16](https://github.com/microsoft/rnx-kit/releases/tag/%40rnx-kit%2Fmetro-config%401.3.16), `@rnx-kit/metro-config` should work out of box with Expo:

```js
const { makeMetroConfig } = require("@rnx-kit/metro-config");
const { getDefaultConfig } = require("expo/metro-config");

const config = getDefaultConfig(__dirname);
module.exports = makeMetroConfig(config);
```

**Comment by iMuhammadMustafa** [comment]
I'm trying to follow your instruction to an existing expo project but I can't seem to be able to get this to work.

the major differences in my project from yours is: 

index.js (since expo doesn't have App.js, I'm using the index page)
```js
import { registerRootComponent } from "expo";
import Index from "./src/app/index.tsx";
registerRootComponent(Index);
```

metro.config.js (because I'm using NativeWind, I'm wrapping your metro's export )
```js
module.exports = withNativeWind(
  {
    ...config,
    //Rest of your code
   //...
  }, 
  { input: "./src/app/global.css", configPath: "./tailwind.config.js" },
);
```

I'm getting this error when trying to run yarn windows
```ps1
>yarn windows
yarn run v1.22.19
$ react-native run-windows --logging --arch arm64
 i Running x64 node on a x64 machine
 i Verbose: ON
 × Solution file is required but not specified in react-native.config. Optionally, use --sln {slnFile}.
Looking for VS installs with version range: [17.11.0,18.0)
Looking for vswhere at: C:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe
 ‼ No public VS release found
 i Trying pre-release VS
Looking for VS installs with version range: [17.11.0,18.0]
Looking for vswhere at: C:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe
 × Could not find MSBuild with VCTools for Visual Studio 17.11.0 or later. Make sure all required components have been installed
 × It is possible your installation is missing required software dependencies. Dependencies can be automatically installed by running C:\Home\Projects\Budgeteer\node_modules\react-native-windows\scripts\rnw-dependencies.ps1 from an elevated PowerShell prompt.
For more information, go to http://aka.ms/rnw-deps
Command failed with error NoMSBuild: Could not find MSBuild with VCTools for Visual Studio 17.11.0 or later. Make sure all required components have been installed
Your telemetry sessionId was eef76706f3346a0cfaffb4a6d4eb46f3
error Command failed with exit code 2300.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

I did run `rnw-dependencies.ps1` and it passed, so I'm not sure what's missing. 

btw do I need to run `npx react-native init-windows --overwrite` ?

**Comment by iMuhammadMustafa** [comment]
Hello @shirakaba , 
Thank you for putting together this walkthrough ,I'm trying to follow your instruction to an existing expo project but I can't seem to be able to get this to work.

the major differences in my project from yours is: 

index.js (since expo doesn't have App.js, I'm using the index page)
```
import { registerRootComponent } from "expo";
import { App } from "expo-router/build/qualified-entry";

registerRootComponent(App);
```
I got the `App` from `node_module/expo-router/entry.js` which is the default main in package.json of an expo application
```
// `@expo/metro-runtime` MUST be the first import to ensure Fast Refresh works
// on web.
import '@expo/metro-runtime';

import { App } from 'expo-router/build/qualified-entry';
import { renderRootComponent } from 'expo-router/build/renderRootComponent';

// This file should only import and register the root. No components or exports
// should be added here.
renderRootComponent(App);
```

second change is in `metro.config.js` I had this line so I kept it at the end of your `metro.config.js` without changing anything in it
```
module.exports = withNativeWind(config, { input: "./src/app/global.css", configPath: "./tailwind.config.js" });
```

I also added to `react-native.config.js` the project and solution path

**Comment by mataide** [comment]
@shirakaba I am getting this error:

![Image](https://github.com/user-attachments/assets/426616b7-659c-464e-bebe-1050457c9b6f)

**Comment by shirakaba** [comment]
@acoates-ms Just in case you haven't seen the [custom-prebuild-example](https://github.com/byCedric/custom-prebuild-example/tree/main) repo, the Expo team showed me this as an example of how they develop config plugins outside of the Expo monorepo before figuring out how to upstream them later.

They show some such config plugins in the [app/plugins](https://github.com/byCedric/custom-prebuild-example/tree/main/app/plugins) folder, which even has both `macos` and `windows` directories to begin with.
